snuba/admin/clickhouse/migration_checks.py

```diff
@@ -0,0 +1,215 @@
+
+    @abstractmethod
+    def can_run(self, migration_id: str) -> ResultReason:
+        raise NotImplementedError
+
+    @abstractmethod
+    def can_reverse(self, migration_id: str) -> ResultReason:
+        raise NotImplementedError
+
+
+class StatusChecker(Checker):
+    """
+    The StatusChecker validates whether you can run or
+    reverse a migration based on the statuses of the
+    migration itself and the ones preceeding/following
+    it depending on the check.
+
+    Preceeding and following are based on the migration_id
+    numbering order and assumes that holds.
+    """
+
+    def __init__(
+        self,
+        migration_group: MigrationGroup,
+        migrations: Sequence[MigrationDetails],
+    ) -> None:
+        self.__group = migration_group
+        migration_statuses = {}
+        for migration_id, status, _ in migrations:
+            migration_statuses[migration_id] = {
+                "migration_id": migration_id,
+                "status": status,
+            }
+        self.__migration_statuses = migration_statuses
+
+    @property
+    def all_migration_ids(self) -> Sequence[str]:
+        return list(self.__migration_statuses.keys())
+
+    def can_run(self, migration_id: str) -> ResultReason:
+        """
+        Covers the following cases:
+        * no running a migration that is already pending or completed
+        * no running a migration that has a preceeding migration that is not completed
+        """
+        all_migration_ids = self.all_migration_ids
+        if self.__migration_statuses[migration_id]["status"] != Status.NOT_STARTED:
+            return ResultReason(False, Reason.ALREADY_RUN)
+
+        for m in all_migration_ids[: all_migration_ids.index(migration_id)]:
+            if self.__migration_statuses[m]["status"] != Status.COMPLETED:
+                return ResultReason(False, Reason.NEEDS_EARLIER_MIGRATIONS)
+
+        return ResultReason(True, Reason.NO_REASON_NEEDED)
+
+    def can_reverse(self, migration_id: str) -> ResultReason:
+        """
+        Covers the following cases:
+        * no reversing a migration that is has not started
+        * no reversing a migration that has a subsequent migration that has not been run
+        """
+        all_migration_ids = self.all_migration_ids
+        if self.__migration_statuses[migration_id]["status"] == Status.NOT_STARTED:
+            return ResultReason(False, Reason.NOT_RUN_YET)
+
+        for m in all_migration_ids[all_migration_ids.index(migration_id) + 1 :]:
+            if self.__migration_statuses[m]["status"] != Status.NOT_STARTED:
+                return ResultReason(False, Reason.NEEDS_SUBSEQUENT_MIGRATIONS)
+
+        return ResultReason(True, Reason.NO_REASON_NEEDED)
+
+
+class PolicyChecker(Checker):
+    """
+    The PolicyChecker validates whether you can run or
+    reverse a migration based on the policy for the
+    migration's group.
+
+    Policies are defined in the ADMIN_ALLOWED_MIGRATION_GROUPS
+    setting.
```

Comment:
Does this class need to exist at all ?
It feels we are maintaining state (the policy and the migration group) for no reason, and by doing so you introduce several possible ways of misusing this class (passing a migration_id that does not correspond to the group you initiualized the `PolicyChecker` with.

What about a two simple methods `can_run`, `can_reverse`. They take in the `MigrationKey`, from there they get the group, from there they load the policies and from there they apply them. 
Messing up becomes much harder.